viadot/sources/cloud_for_customers.py

```diff
@@ -0,0 +1,256 @@
+import re
+import requests
+import pandas as pd
+
+from copy import deepcopy
+from typing import Any, Dict, List, Optional
+from pydantic import BaseModel
+from urllib.parse import urljoin
+from viadot.exceptions import CredentialError
+
+from ..utils import handle_api_response
+from ..config import get_source_credentials
+from .base import Source
+
+
+class CloudForCustomersCredentials(BaseModel):
+    username: str  # CloudForCustomers username (e.g username@{tenant_name}.com).
+    password: str  # CloudForCustomers password.
+
+
+class CloudForCustomers(Source):
+    """Cloud for Customers connector to fetch Odata source.
+
+    Args:
+        url (str, optional): The URL to the C4C API. E.g 'https://my336539.crm.ondemand.com/sap/c4c/odata/v1/c4codataapi/'.
+        endpoint (str, optional): The API endpoint.
+        report_url (str, optional): The API url in case of prepared report.
+        filter_params (Dict[str, Any], optional): Filtering parameters passed to the request. E.g {"$filter": "AccountID eq '1234'"}. More info: https://userapps.support.sap.com/sap/support/knowledge/en/2330688
+        credentials (CloudForCustomersCredentials, optional): Cloud for Customers credentials.
+        config_key (str, optional): The key in the viadot config holding relevant credentials.
+    """
+
+    DEFAULT_PARAMS = {"$format": "json"}
+
+    def __init__(
+        self,
+        url: str = None,
+        endpoint: str = None,
+        report_url: str = None,
+        filter_params: Dict[str, Any] = None,
+        credentials: CloudForCustomersCredentials = None,
+        config_key: Optional[str] = None,
+        *args,
+        **kwargs,
+    ):
+
+        ## Credentials logic
+        credentials = credentials or get_source_credentials(config_key)
+        if credentials is None:
+            raise CredentialError("Please specify the credentials.")
+        CloudForCustomersCredentials(**credentials)  # validate the credentials schema
+        super().__init__(*args, credentials=credentials, **kwargs)
+        ## End Credentials logic
+
+        self.url = url or self.credentials.get("server")
+        self.report_url = report_url
+
+        if self.url is None and report_url is None:
+            raise CredentialError("One of: ('url', 'report_url') is required.")
+
+        self.is_report = bool(report_url)
+        self.query_endpoint = endpoint
+
+        if filter_params:
+            filter_params_merged = self.DEFAULT_PARAMS.copy()
+            filter_params_merged.update(filter_params)
+
+            self.filter_params = filter_params_merged
+        else:
+            self.filter_params = self.DEFAULT_PARAMS
+
+        if self.url:
+            self.full_url = urljoin(self.url, self.query_endpoint)
+
+    @staticmethod
+    def create_metadata_url(url: str) -> str:
+        """Creates url to fetch metadata to.
```

Comment:
Did you mean from? Also I don't understand this explanation, it feels like half of the information.

Reply:
Yes, typo, I meant "from".
Fixed